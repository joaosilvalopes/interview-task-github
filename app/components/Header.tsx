import { Link } from 'react-router-dom';
import styled from 'styled-components';

import useTheme from '~/hooks/useTheme';

const HeaderContainer = styled.header`
    display: flex;
    justify-content: space-between;
    height: ${props => props.theme.headerHeight};
    background-color: ${props => props.theme.colors.headerBackground};
    color: ${props => props.theme.colors.header};
    font-size: 2rem;
    padding: 0 2rem;
`;

const HomeLink = styled(Link)`
    text-decoration: none;
    color: currentColor;

    &:focus {
        color: currentColor;
    }

    &:hover {
        color: ${props => props.theme.colors.homeLinkOnHover};
    }
`;

const ChangeThemeButton = styled.button`
    border: 0;
    background-color: transparent;
    cursor: pointer;

    &:hover {
        color: ${props => props.theme.colors.changeThemeButtonOnHover};
    }
`;

const Header = () => {
    const [theme,setTheme] = useTheme();

    return (
        <HeaderContainer>
            <HomeLink data-testid="home-link" to="/">⌂</HomeLink>
            <ChangeThemeButton data-testid="change-theme-button" onClick={() => setTheme(prevTheme => prevTheme === 'light' ? 'dark' : 'light')}>{theme === 'light' ? '☀' : '☽'}</ChangeThemeButton>
        </HeaderContainer>
    )
};

export default Header;
