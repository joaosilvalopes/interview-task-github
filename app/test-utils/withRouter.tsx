import { useState, useEffect, ComponentType } from "react";
import { To, Router, Routes, Route } from "react-router";
import { History } from "@remix-run/router";

const withRouter = (Component: ComponentType, { history, path }: { history: History, path: string }) => () => {
    const [location, setLocation] = useState(history.location);

    useEffect(() => {
        history.listen(({ location }) => setLocation(location));
    }, []);

    const push = (to: To) => setTimeout(() => history.push(to), 100); // simulate remix loader fetch before changing route

    return <Router location={location} navigator={{ ...history, push }}>
        <Routes>
            <Route path={path} element={<Component />} />
            <Route path="*" element={null} />
        </Routes>
    </Router>;
};

export default withRouter;
