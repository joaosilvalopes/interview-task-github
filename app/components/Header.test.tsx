import { createMemoryHistory } from '@remix-run/router';
import { render, fireEvent, waitFor } from '@testing-library/react';
import withRouter from '~/test-utils/withRouter';

import ThemeProvider from "~/theme";

import Header from './Header';

const history = createMemoryHistory({ v5Compat: true });

const HeaderWithRouter = withRouter(Header, { path: '/', history });

describe('Header component', () => {
  it('renders correctly and changes theme on change theme button click', async () => {
    const { getByTestId, asFragment } = render(<ThemeProvider><HeaderWithRouter /></ThemeProvider>);

    expect(getByTestId('change-theme-button')).toBeInTheDocument();
    expect(getByTestId('home-link')).toBeInTheDocument();
    expect(getByTestId('change-theme-button')).toHaveTextContent('☀');

    fireEvent.click(getByTestId('change-theme-button'));

    expect(getByTestId('change-theme-button')).toHaveTextContent('☽');
    expect(asFragment()).toMatchSnapshot();
  });

  it('Redirects to / on home link click', async () => {
    const { getByTestId, asFragment } = render(<ThemeProvider><HeaderWithRouter /></ThemeProvider>);

    expect(getByTestId('change-theme-button')).toBeInTheDocument();
    expect(getByTestId('home-link')).toBeInTheDocument();

    fireEvent.click(getByTestId('home-link'));

    await waitFor(() => {
      expect(history.location.pathname).toBe(`/`);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});

