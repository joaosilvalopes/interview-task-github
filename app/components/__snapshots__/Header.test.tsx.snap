// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Redirects to / on home link click 1`] = `
<DocumentFragment>
  <header
    class="sc-beySbM fiMPnF"
  >
    <a
      class="sc-guDLRT hbtjoA"
      data-testid="home-link"
      href="/"
    >
      ⌂
    </a>
    <button
      class="sc-dmyDGi hpQBLb"
      data-testid="change-theme-button"
    >
      ☀
    </button>
  </header>
</DocumentFragment>
`;

exports[`Header component renders correctly and changes theme on change theme button click 1`] = `
<DocumentFragment>
  <header
    class="sc-beySbM ezJJWJ"
  >
    <a
      class="sc-guDLRT hLzZhb"
      data-testid="home-link"
      href="/"
    >
      ⌂
    </a>
    <button
      class="sc-dmyDGi jMgBXg"
      data-testid="change-theme-button"
    >
      ☽
    </button>
  </header>
</DocumentFragment>
`;
